lib/util/src/workqueue.c

```diff
@@ -7,88 +7,161 @@
- * @wq_base:        base of workqueue memory allocation for free()
  * @wq_priv:        flexible array of worker thread private data structs
  */
 struct workqueue_struct {
-    struct mutex     wq_lock;
-    struct list_head wq_pending;
-    bool             wq_shutdown;
-    int              wq_tdcnt;
-    struct cv        wq_idle;
-    struct cv        wq_barrier;
-    int              wq_refcnt;
-    uint             wq_barid;
-    int              wq_tdmax;
-    struct list_head wq_delayed;
-    char             wq_name[16];
-    void *           wq_base;
-    struct wq_priv   wq_priv[];
+    struct mutex      wq_lock;
+    struct list_head  wq_pending;
+    bool              wq_running;
+    bool              wq_growing;
+    int               wq_refcnt;
+    int               wq_tdcnt;
+    int               wq_tdmax;
+    int               wq_tdmin;
+    uint              wq_barid;
+    uint              wq_tcdelay;
+    struct cv         wq_barrier;
+    struct cv         wq_idle;
+    struct list_head  wq_delayed;
+    struct timer_list wq_grow;
+    char              wq_name[16];
+    struct wq_priv    wq_priv[];
 };
 
 static void
-workqueue_dump(struct workqueue_struct *wq);
-static void
-flush_barrier(struct work_struct *work);
+dump_workqueue_locked(struct workqueue_struct *wq);
+
 static void *
 worker_thread(void *arg);
 
+static HSE_ALWAYS_INLINE struct work_struct *
+workqueue_first(struct workqueue_struct *wq)
+{
+    return list_first_entry_or_null(&wq->wq_pending, struct work_struct, entry);
+}
+
+static void
+grow_workqueue_cb(ulong arg)
+{
+    struct workqueue_struct *wq = (void *)arg;
+    struct wq_priv *priv = NULL;
+    int rc, i;
+
+    mutex_lock(&wq->wq_lock);
+    for (i = 0; i < wq->wq_tdmax; ++i) {
+        priv = wq->wq_priv + (wq->wq_tdcnt + i) % wq->wq_tdmax;
+        if (!priv->wqp_wq)
+            break;
+    }
+
+    assert(priv && i < wq->wq_tdmax);
+
+    priv->wqp_wq = wq;
+    ++wq->wq_refcnt;
+    ++wq->wq_tdcnt;
+    mutex_unlock(&wq->wq_lock);
+
+    rc = pthread_create(&priv->wqp_tid, NULL, worker_thread, priv);
+    if (rc) {
+        merr_t err = merr(rc);
+
+        log_warnx("growing failed (%s %d/%d%d): @@e",
+                  err, wq->wq_name, wq->wq_tdcnt, wq->wq_tdmin, wq->wq_tdmax);
+    }
+
+    mutex_lock(&wq->wq_lock);
+    if (rc) {
+        struct work_struct *first = workqueue_first(wq);
+
+        /* If there is a flush in progress then we must awaken all
+         * visitors so that each can re-evaluate the barrier state.
+         */
+        if (first && !first->func)
+            cv_broadcast(&wq->wq_barrier);
+
+        priv->wqp_wq = NULL;
+        --wq->wq_refcnt;
+        --wq->wq_tdcnt;
+    }
+
+    wq->wq_growing = workqueue_first(wq) && (wq->wq_tdcnt < wq->wq_tdmax);
+    if (wq->wq_growing) {
+        wq->wq_grow.expires = jiffies + wq->wq_tcdelay;
+        add_timer(&wq->wq_grow);
+        ++wq->wq_refcnt;
+    }
+
+    --wq->wq_refcnt;
```

Comment:
This decrement seems off, probably deserves a comment.

Reply:
Maybe it decrements the ref that was taken prior to queueing this cb?

Reply:
Agreed, this is dropping the reference taken by queue_work()...

Reply:
I just updated it, hope it's more clear!

Reply:
Thanks! When called in the context of alloc_workqueue(), wq_refcnt is initialized to min_active and that's decremented here, right?

Reply:
Good eye!  Yeah, that's a startup condition so that I could reuse this function to grow the initial threads vs having another chunk of code in alloc-workqueue calling pthread-create.

Reply:
Nice, I liked that!